File: frontend/packages/employee-frontend/src/api/child/backup-care.ts
// SPDX-FileCopyrightText: 2017-2020 City of Espoo
//
// SPDX-License-Identifier: LGPL-2.1-or-later

import { UUID } from '~types'
import { Failure, Result, Success } from '~api/index'
import { ChildBackupCare, UnitBackupCare } from '~types/child'
import { client } from '~api/client'
import { JsonOf } from '@evaka/lib-common/src/json'
import LocalDate from '@evaka/lib-common/src/local-date'

interface BackupCaresResponse<T extends ChildBackupCare | UnitBackupCare> {
  backupCares: T[]
}

export async function getChildBackupCares(
  childId: UUID
): Promise<Result<ChildBackupCare[]>> {
  return client
    .get<JsonOf<BackupCaresResponse<ChildBackupCare>>>(
      `/children/${childId}/backup-cares`
    )
    .then((res) =>
      Success.of(
        res.data.backupCares.map((x) => ({
          ...x,
          period: {
            start: LocalDate.parseIso(x.period.start),
            end: LocalDate.parseIso(x.period.end)
          }
        }))
      )
    )
    .catch((e) => Failure.fromError(e))
}

interface NewBackupCare {
  unitId: UUID
  groupId?: UUID
  period: {
    start: LocalDate
    end: LocalDate
  }
}

interface BackupCareCreateResponse {
  id: UUID
}

export async function createBackupCare(
  childId: UUID,
  payload: NewBackupCare
): Promise<Result<UUID>> {
  return client
    .post<JsonOf<BackupCareCreateResponse>>(
      `/children/${childId}/backup-cares`,
      payload
    )
    .then((res) => Success.of(res.data.id))
    .catch((e) => Failure.fromError(e))
}

interface BackupCareUpdateRequest {
  period: {
    start: LocalDate
    end: LocalDate
  }
  groupId?: UUID
}

export async function updateBackupCare(
  backupCareId: UUID,
  payload: BackupCareUpdateRequest
): Promise<Result<void>> {
  return client
    .post(`/backup-cares/${backupCareId}`, payload)
    .then(() => Success.of(undefined))
    .catch((e) => Failure.fromError(e))
}

export async function removeBackupCare(
  backupCareId: UUID
): Promise<Result<void>> {
  return client
    .delete(`/backup-cares/${backupCareId}`)
    .then(() => Success.of(undefined))
    .catch((e) => Failure.fromError(e))
}

File: service/src/main/kotlin/fi/espoo/evaka/backupcare/BackupCare.kt
// SPDX-FileCopyrightText: 2017-2020 City of Espoo
//
// SPDX-License-Identifier: LGPL-2.1-or-later

package fi.espoo.evaka.backupcare

import fi.espoo.evaka.serviceneed.ServiceNeed
import fi.espoo.evaka.shared.BackupCareId
import fi.espoo.evaka.shared.ChildId
import fi.espoo.evaka.shared.DaycareId
import fi.espoo.evaka.shared.GroupId
import fi.espoo.evaka.shared.domain.FiniteDateRange
import fi.espoo.evaka.shared.security.Action
import java.time.LocalDate
import org.jdbi.v3.core.mapper.Nested
import org.jdbi.v3.core.mapper.PropagateNull
import org.jdbi.v3.json.Json

data class ChildBackupCare(
    val id: BackupCareId,
    @Nested("unit_") val unit: BackupCareUnit,
    @Nested("group_") val group: BackupCareGroup?,
    val period: FiniteDateRange
)

data class ChildBackupCareResponse(
    val backupCare: ChildBackupCare,
    val permittedActions: Set<Action.BackupCare>
)

data class UnitBackupCare(
    val id: BackupCareId,
    @Nested("child_") val child: BackupCareChild,
    @Nested("group_") val group: BackupCareGroup?,
    val period: FiniteDateRange,
    @Json val fromUnits: List<String>,
    @Json val serviceNeeds: Set<ServiceNeed>,
    val missingServiceNeedDays: Int
)

data class GroupBackupCare(
    val id: BackupCareId,
    val childId: ChildId,
    val period: FiniteDateRange
)

data class BackupCareChild(
    val id: ChildId,
    val firstName: String,
    val lastName: String,
    val birthDate: LocalDate
)

data class BackupCareUnit(
    val id: DaycareId,
    val name: String
)

data class BackupCareGroup(
    @PropagateNull val id: GroupId,
    val name: String
)

data class NewBackupCare(
    val unitId: DaycareId,
    val groupId: GroupId?,
    val period: FiniteDateRange
)
